### PR TITLE
Allow support to pass function in the context

### PR DESCRIPTION
**Pull request by pranavwani**
Description: Useful, for applying the function in the payload.

Usecase:
```
const parse = require("json-templates");
const faker = require("faker");

const context = {
   $name: faker.name.findName,
   $email: faker.internet.email
}

const content = `{
  "name": "{{$name}}"   
}`

const template = parse(content);

const payload = template({ ...context });
```

Output:
```
{
  "name": "Rowan Nikolaus"
}
```

### CODE REVIEW

**Comment by curran** [comment]
Interesting...

I'm curious, why not do it like this?

```js
const parse = require("json-templates");
const faker = require("faker");

const context = {
   $name: faker.name.findName(),  // <-------- This changed
   $email: faker.internet.email
}

const content = `{
  "name": "{{$name}}"   
}`

const template = parse(content);

const payload = template({ ...context });
```

**Comment by pranavwani** [comment]
@curran for generating dynamic value needs to assign a prototype rather than calling a function because by calling a function it will initialize with the static value.

Requirement: Single variable call multiple times at runtime.

**Output:**
By assigning function call
```
{
  "name": "Eva Quigley"   
}
{
  "name": "Eva Quigley"   
}

```

**Comment by curran** [comment]
Interesting...

I'm curious why not do this?

```js
setInterval(() => {
  const context = {
    $name: faker.name.findName(), // <-------- This changed
    $email: faker.internet.email,
  };

  const content = `{
    "name": "{{$name}}"   
  }`;

  const template = parse(content);

  const payload = template({ ...context });
  console.log(payload);
}, 2000);
```

**Comment by curran** [comment]
I'm open to accept the PR, just want to understand the use case more deeply.

Please add a test case that covers this new feature. Thanks!

**Comment by pranavwani** [comment]
@curran test case is added, two dependencies were added to full fill the use case.

File: test.js [hunk 1]
```diff
@@ -305,6 +363,33 @@ describe('json-template', () => {
...
+
+    const userCard = {
+      uuid: '{{$uuidV4}}',
+      name: { firstName: '{{$firstName}}' },
+      city: '{{city}}',
+    };
+
+    const result = applyVariables(userCard, { city: 'Surat' });
+
+    it('should compute template with Function', function () {
```
Comment (curran) [inline test.js:379]: This test is overly complex. The usage of uuid is not required. You can mock it out to return a fixed value.

The test should only test the feature, not additional things.

**Review by curran — CHANGES_REQUESTED** [review]
I appreciate your work here, but I see several issues with this PR:

 * It changes existing dependency versions, which is not required.
 * It adds new dependencies that are not required.
 * The test is overly complex, no need to use the UID function itself.
 * There are incidental formatting changes.

What I'd like to see in a PR is a _very simple_ test that only serves the purpose of verifying the new code path. No need to introduce additional dependencies.

Please, it might be easiest to start a new PR that only includes the feature implementation and a single new test. No formatting changes, no dependency changes.

Thanks!

**Comment by pranavwani** [comment]
@curran sorry for the delay. I'll close this PR because of the code format issues and create the new one by adding the test case without any dependencies.

File: package.json [hunk 2]
```diff
@@ -26,11 +26,15 @@
   "homepage": "https://github.com/datavis-tech/json-templates#readme",
   "devDependencies": {
     "mocha": "^8.1.3",
-    "rollup": "^2.27.0",
-    "rollup-plugin-buble": "^0.19.8"
+    "rollup": "^0.65.2",
```
[the remark this answers is not in the document]
Comment (pranavwani) [reply package.json:29]: I think yarn auto-updates the libs.

File: package.json [hunk 3]
```diff
@@ -26,11 +26,15 @@
...
+    "rollup-plugin-buble": "^0.19.2"
   },
   "dependencies": {
     "dedupe": "^3.0.2",
-    "object-path": "^0.11.4"
+    "object-path": "^0.11.4",
+    "faker": "^5.1.0",
```
[the remark this answers is not in the document]
Comment (pranavwani) [reply package.json:35]: I'll remove them

File: test.js [hunk 4]
```diff
@@ -39,8 +67,8 @@ describe('json-template', () => {
       assert.deepEqual(template.parameters, [
         {
           key: 'foo',
-          defaultValue: 'bar'
-        }
+          defaultValue: 'bar',
```
[the remark this answers is not in the document]
Comment (pranavwani) [reply test.js:70]: :+1:

File: test.js [hunk 5]
```diff
@@ -86,38 +122,47 @@ describe('json-template', () => {
...
+        { key: 'startTime', defaultValue: 'now-24h' },
+      ]);
       assert.equal(template({ startTime: 'now-48h' }), 'now-48h');
       assert.equal(template(), 'now-24h');
     });
 
     it('should handle special characters in defaults', () => {
       const template = parse('{{foo:-+., @/()?=*_}}');
-      assert.deepEqual(template.parameters, [{ key: 'foo', defaultValue: '-+., @/()?=*_' }]);
+      assert.deepEqual(template.parameters, [
```
[the remark this answers is not in the document]
Comment (pranavwani) [reply test.js:134]: Ya, I've used Prettier. But I'll remove it.

File: test.js (hunk 1 above)
Comment (pranavwani) [reply test.js:379]: I'll keep it simple.

**Comment by curran** [comment]
Perfect! Thank you 🙏

**Pull request closed without merging**